core/charm/strategies.go

```diff
@@ -0,0 +1,320 @@
+		return nil, errors.Trace(err)
+	}
+	return &Strategy{
+		charmURL: curl,
+		store: StoreCharmStore{
+			charmRepo: charmRepo,
+		},
+		force: force,
+	}, nil
+}
+
+// DownloadFromCharmHub will creates a procedure to install a charm from the
+// charm hub.
+func DownloadFromCharmHub(url string, force bool) (*Strategy, error) {
+	curl, err := charm.ParseURL(url)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &Strategy{
+		charmURL: curl,
+		store:    StoreCharmHub{},
+	}, nil
+}
+
+// CharmURL returns the strategy URL associated with it.
+func (p *Strategy) CharmURL() *charm.URL {
+	return p.charmURL
+}
+
+// Validate will attempt to validate the requirements for adding a charm to the
+// store.
+func (p *Strategy) Validate() error {
+	if err := p.store.Validate(p.charmURL); err != nil {
+		return errors.Trace(err)
+	}
+	if p.charmURL.Revision < 0 {
+		return errors.Errorf("charm URL must include a revision")
+	}
+	return nil
+}
+
+// DownloadResult defines the result from the attempt to install a charm into
+// state.
+type DownloadResult struct {
+	// Charm is the metadata about the charm for the archive.
+	Charm StoreCharm
+
+	// Data contains the bytes of the archive.
+	Data io.Reader
+
+	// Size is the number of bytes in Data.
+	Size int64
+
+	// SHA256 is the hash of the bytes in Data.
+	SHA256 string
+}
+
+// Run the procedure against the correct store.
+func (p *Strategy) Run(state State, version VersionValidator) (DownloadResult, bool, error) {
+	charm, err := state.PrepareCharmUpload(p.charmURL)
+	if err != nil {
+		return DownloadResult{}, false, errors.Trace(err)
+	}
+
+	// Charm is already in state, so we can exit out early.
+	if charm.IsUploaded() {
+		return DownloadResult{}, true, nil
+	}
+
+	// Get the charm and its information from the store.
+	file, err := ioutil.TempFile("", p.charmURL.Name)
+	if err != nil {
+		return DownloadResult{}, false, errors.Trace(err)
+	}
+
+	p.deferFunc(func() error {
+		_ = file.Close()
+		return nil
+	})
+	p.deferFunc(func() error {
+		_ = os.Remove(file.Name())
+		return nil
+	})
+
+	archive, checksum, err := p.store.Download(p.charmURL, file.Name())
+	if err != nil {
+		return DownloadResult{}, false, errors.Trace(err)
+	}
+
+	if err := version.Validate(archive.Meta()); err != nil {
+		return DownloadResult{}, false, errors.Trace(err)
+	}
+
+	// Validate the charm lxd profile once we've downloaded it.
+	if err := lxdprofile.ValidateLXDProfile(makeStoreCharmLXDProfiler(archive)); err != nil && !p.force {
+		return DownloadResult{}, false, errors.Annotate(err, "cannot add charm")
+	}
+
+	result, err := p.downloadResult(file.Name(), checksum)
+	if err != nil {
+		return DownloadResult{}, false, errors.Trace(err)
+	}
+
+	return DownloadResult{
+		Charm:  archive,
+		Data:   result.Data,
+		Size:   result.Size,
+		SHA256: result.SHA256,
+	}, false, nil
+}
+
+// Finish will attempt to close out the procedure and clean up any outstanding
+// tasks.
+func (p *Strategy) Finish() error {
```

Comment:
In the same way, we generally do in most of the code of go. Essentially the code offers an error but might use it or not! Generally, we should handle it, but the previous code of `AddCharmWithAuthorizationAndRepo` just ignored it within a defer.  Guess it might be good practice eventually to handle the errors, even just outputting to a log.